pkg/finalize_bundle.go

```diff
@@ -0,0 +1,188 @@
+package cnab_riff
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/deislabs/cnab-go/bundle"
+	"github.com/deislabs/duffle/pkg/duffle/manifest"
+	"github.com/ghodss/yaml"
+	"github.com/pivotal/go-ape/pkg/furl"
+	"github.com/pivotal/image-relocation/pkg/image"
+	"github.com/pivotal/image-relocation/pkg/registry"
+	"github.com/projectriff/cnab-k8s-installer-base/pkg/apis/kab/v1alpha1"
+	"github.com/projectriff/k8s-manifest-scanner/pkg/scan"
+)
+
+// this performs following tasks:
+// 1. inlines the content of the resource url into the bundle
+// 2. adds images to duffle.json by scanning the resource content
+// 3. computes digests for images
+// 4. replaces image references in kab manifest with digested references
+func FinalizeBundle(bundlePath, kabManifestPath string) error {
+	mfst := &manifest.Manifest{}
+	err := unmarshallFile(bundlePath, mfst)
+	if err != nil {
+		return err
+	}
+
+	err = InlineContentInKabManifest(kabManifestPath)
+	if err != nil {
+		return err
+	}
+
+	images, err := GetImagesFromKabManifest(kabManifestPath)
+	if err != nil {
+		return err
+	}
+
+	mfst.Images = map[string]bundle.Image{}
+	r := registry.NewRegistryClient()
+	replacements := []string{}
+
+	for _, img := range images {
+		name, err := image.NewName(img)
```

Comment:
I find this loop difficult to follow. Using more descriptive names and defining variables closer to where they are used could go a long way.

Reply:
Maybe factoring out some of the logic inside the body of the loop into helper methods would improve the readability.